### PR TITLE
PWN-7343 - get prices on Ethereum tokens and tests on it

### PR DESCRIPTION
**Pull request by kabramovp2p**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7343

## Description of Work

PWN-7343 - get prices on Ethereum tokens and tests on it

### CODE REVIEW

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/user/repository/prices/impl/TokenPricesCoinGeckoRepository.kt [hunk 1]
```diff
@@ -36,4 +37,30 @@ class TokenPricesCoinGeckoRepository(
...
+        tokenAddresses: List<TokenAddress>,
+        targetCurrencySymbol: String
+    ): List<TokenPrice> =
+        withContext(dispatchers.io) {
+            val tokenAddressesForRequest = tokenAddresses.joinToString(",") { it.address }
+            coinGeckoApi.getEthereumTokenPrices(
+                tokenAddresses = tokenAddressesForRequest,
+                targetCurrency = targetCurrencySymbol.lowercase()
+            ).map { (tokenAddress, tokenPrice) ->
+                TokenPrice(tokenId = tokenAddress, price = tokenPrice.currentPrice)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/repository/prices/impl/TokenPricesCoinGeckoRepository.kt:63]: we can be confused later when using TokenPrice class
everywhere for SOL tokenId for us is coingeckoId or some id, not address
maybe create own wrapper class for Eth prices?

File: app/src/test/java/org/p2p/wallet/utils/TokenBalancesAndPricesTest.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                contractAddresses = tokenAddress
+            )
+            val tokenMetadata = TokenConverter.fromNetwork(tokenAddress.toString(), tokenMetadataResponse)
+            val tokenPrice = tokenPriceMap[TokenAddress(tokenAddress.toString())]
+
+            stringBuilder.append("Token Metadata: $tokenMetadata ")
+            stringBuilder.append("Token Price: $tokenPrice ")
+            stringBuilder.append("\n")
+        }
+        println(stringBuilder)
```
Comment (gslevinkov) [inline app/src/test/java/org/p2p/wallet/utils/TokenBalancesAndPricesTest.kt:94]: test doesn't check anything, no asserts, only println
don't forger to remove println

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/balance/BalanceRepository.kt [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
 
-import com.google.gson.JsonElement
+import java.math.BigInteger
 import org.p2p.ethereumkit.external.api.response.TokenBalancesResponse
 import org.p2p.ethereumkit.external.api.response.TokenMetadataResponse
 import org.p2p.ethereumkit.internal.models.EthAddress
-import java.math.BigInteger
 
-internal interface BalanceRepository {
+interface BalanceRepository {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/balance/BalanceRepository.kt:8]: i think you are getting balance through EthereumKitRepository, why do you remove internal?
if it's for tests - just write tests in this module, this is not a reason to break incapsulation :(

File: ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt [hunk 4]
```diff
@@ -47,11 +47,13 @@ class TokenBalancesTest : KoinTest {
...
             val tokenBalance = balance.balances.first { it.contractAddress == address }.tokenBalance
             stringBuilder.append("Token Name: $tokenName ")
+            stringBuilder.append("Token Address: $address ")
             stringBuilder.append("Token Symbol: $tokenSymbol ")
             stringBuilder.append("Token Balance: $tokenBalance ")
             stringBuilder.append("Token Decimals: $tokenDecimals ")
             stringBuilder.append("Token logo: $tokenLogo ")
             stringBuilder.append("\n")
         }
+        println(stringBuilder)
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt:57]: don't forget to remove

Comment (gslevinkov) [reply ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt:57]: if you need a help to write tests with asserts - please come to DM

**Review by gslevinkov — CHANGES_REQUESTED** [review]

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt [hunk 5]
```diff
@@ -1,28 +1,36 @@
...
                 alchemyService = get(),
                 networkEnvironment = EthereumNetworkEnvironment.ALCHEMY,
                 gson = get()
             )
         }
+        single<PriceRepository> {
+            EthereumPriceRepository(priceApi = get())
+        }
+
+        singleOf<CoroutineDispatchers> { DefaultDispatchers() }
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt:34]: i think this dependency would be visible in global dependency graph

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/EthereumNetworkModule.kt [hunk 6]
```diff
@@ -17,12 +19,14 @@ import java.util.logging.Logger
...
 
-object EthereumNetworkModule {
+internal object EthereumNetworkModule {
 
     fun create(): Module = module {
         single(named(QUALIFIER_ETH_HTTP_CLIENT)) { getOkHttpClient() }
-        single { getRetrofit(get(named(QUALIFIER_ETH_HTTP_CLIENT)),get()) }
-        single { GsonProvider().provide()}
+        single { getRetrofit(get(named(QUALIFIER_ETH_HTTP_CLIENT)), get()) }
+        single { GsonProvider().provide() }
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/EthereumNetworkModule.kt:27]: better to extract this Gson dependency to some Infrastructure module

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/balance/EthereumTokensRepository.kt [hunk 7]
```diff
@@ -1,24 +1,23 @@
...
 import java.math.BigInteger
 import java.net.URI
 
-internal class EthereumBalanceRepository(
+internal class EthereumTokensRepository(
     private val alchemyService: AlchemyService,
     private val networkEnvironment: EthereumNetworkEnvironment,
     private val gson: Gson
-) : BalanceRepository {
+) : TokensRepository {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/balance/EthereumTokensRepository.kt:20]: EthereumTokensRepository
impl should be called EthereumTokensRemoteRepository

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/balance/EthereumTokensRepository.kt [hunk 8]
```diff
@@ -33,9 +32,13 @@ internal class EthereumBalanceRepository(
         return request.parseResponse(response, gson)
     }
 
-    override suspend fun getTokenBalances(address: EthAddress): TokenBalancesResponse {
+    override suspend fun getTokenBalances(
+        address: EthAddress,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/balance/EthereumTokensRepository.kt:36]: address -> userAddress?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt (hunk 5 above)
Comment (gslevinkov) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt:34]: ah, i see
maybe you can extract DefaultDispatchers and CoroutineDispatchers to :core and reuse them here?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/ERC20Tokens.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package org.p2p.ethereumkit.external.model
+
+enum class ERC20Tokens(val contractAddress: String) {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/ERC20Tokens.kt:3]: Erc20Tokens

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt [hunk 10]
```diff
@@ -1,18 +1,9 @@
...
-    private val seedPhrase: List<String>
-) {
-
-    private val mnemonic: ByteArray = Mnemonic().toSeed(seedPhrase)
-
-    val address: EthAddress = Signer.address(mnemonic, Chain.Ethereum)
-
-    val privateKey: BigInteger = Signer.privateKey(mnemonic, Chain.Ethereum)
-}
+data class EthTokenKeyProvider(
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt:6]: Not provider anymore - looks like EthTokenKeysHolder

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenMetadata.kt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package org.p2p.ethereumkit.external.model
+
+import org.p2p.ethereumkit.internal.models.EthAddress
+import java.math.BigDecimal
+import java.math.BigInteger
+
+data class EthTokenMetadata(
+    val contractAddress: EthAddress,
+    val balance: BigInteger,
+    val decimals: Int,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenMetadata.kt:10]: maybe you can create method for getting balance in lamports?

balanceInLamports = balance.fromLamports(decimals)

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenMetadata.kt [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+    val logoUrl: String,
+    val tokenName: String,
+    val symbol: String,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenMetadata.kt:13]: tokenSymbol

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/price/EthereumPriceRepository.kt [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import org.p2p.ethereumkit.external.api.coingecko.CoinGeckoService
+import org.p2p.ethereumkit.external.api.coingecko.response.CoinGeckoPriceResponse
+
+internal class EthereumPriceRepository(
+    private val priceApi: CoinGeckoService
+): PriceRepository {
+
+
+    override suspend fun getTokenPrice(
+        tokenAddresses: List<String>,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/price/EthereumPriceRepository.kt:12]: String -> EthAddress ?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 14]
```diff
@@ -1,20 +1,71 @@
...
+    private val priceRepository: PriceRepository,
+    private val dispatchers: CoroutineDispatchers
 ) : EthereumRepository {
 
-    init {
-        EthereumKit.init()
+    private var tokenKeyProvider: EthTokenKeyProvider? = null
+
+    override suspend fun init(seedPhrase: List<String>) {
+        tokenKeyProvider = EthTokenKeyProvider(
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:27]: maybe you can inject EthTokenKeyProvider here?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 15]
```diff
@@ -1,20 +1,71 @@
...
+        return balanceRepository.getWalletBalance(publicKey)
+    }
+
+    override suspend fun loadWalletTokens(): List<EthTokenMetadata> = withContext(dispatchers.io) {
+        val walletTokens = loadTokensMetadata()
+        val tokensPrice = getPriceForTokens(tokenAddresses = walletTokens.map { it.contractAddress.toString() })
+        tokensPrice.forEach { (address, price) ->
+            walletTokens.find { it.contractAddress.hex == address }?.price = price
+        }
+        return@withContext walletTokens
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:44]: no need for return

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 16]
```diff
@@ -1,20 +1,71 @@
...
+        tokenKeyProvider = EthTokenKeyProvider(
+            publicKey = Signer.address(words = seedPhrase, chain = Chain.Ethereum),
+            privateKey = Signer.privateKey(words = seedPhrase, chain = Chain.Ethereum)
+        )
     }
 
     override suspend fun getBalance(): BigInteger {
-        return balanceRepository.getWalletBalance(tokenKeyProvider.address)
+        val publicKey = tokenKeyProvider?.publicKey ?: error("init function wasn't called")
+        return balanceRepository.getWalletBalance(publicKey)
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:35]: you forgot withContext

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 17]
```diff
@@ -1,20 +1,71 @@
...
+    }
+
+    private suspend fun getPriceForTokens(tokenAddresses: List<String>): Map<String, BigDecimal> {
+        return priceRepository.getTokenPrice(tokenAddresses = tokenAddresses)
+            .mapValues { it.value.priceInUsd }
+    }
+
+    private suspend fun loadTokensMetadata(): List<EthTokenMetadata> = withContext(dispatchers.io) {
+        val publicKey = tokenKeyProvider?.publicKey ?: error("init function wasn't called")
+        val tokenAddresses = ERC20Tokens.values().map { EthAddress(it.contractAddress) }
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:54]: maybe extract this mapping as a method for ERC20Tokens?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 18]
```diff
@@ -1,20 +1,71 @@
...
+        return@withContext balanceRepository.getTokenBalances(address = publicKey, tokenAddresses = tokenAddresses)
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:55]: no need for return

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 19]
```diff
@@ -1,20 +1,71 @@
...
+            .balances
+            .map { token ->
+                async {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:58]: i suggest exracting this map to a method to keep call chain clean

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt [hunk 20]
```diff
@@ -1,20 +1,71 @@
...
+                    EthTokenMetadata(
+                        contractAddress = token.contractAddress,
+                        balance = token.tokenBalance,
+                        decimals = metadata.decimals,
+                        logoUrl = metadata.logoUrl.orEmpty(),
+                        tokenName = metadata.tokenName.orEmpty(),
+                        symbol = metadata.symbol.orEmpty()
+                    )
+                }
+            }.awaitAll()
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/repository/EthereumKitRepository.kt:69]:
```suggestion
            }
            .awaitAll()
```

File: ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt [hunk 21]
```diff
@@ -26,34 +26,21 @@ class TokenBalancesTest : KoinTest {
...
-            stringBuilder.append("Token Symbol: $tokenSymbol ")
-            stringBuilder.append("Token Balance: $tokenBalance ")
-            stringBuilder.append("Token Decimals: $tokenDecimals ")
-            stringBuilder.append("Token logo: $tokenLogo ")
-            stringBuilder.append("\n")
-        }
-        println(stringBuilder)
+        repository.init(seedPhrase = "apart approve black comfort steel spin real renew tone primary key cherry".split(" "))
+        val result = repository.loadWalletTokens()
+        assert(result.isNotEmpty())
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt:43]: `org.assertj.core.api.Assertions.assertThat(result).isNotEmpty`

**Review by gslevinkov — APPROVED** [review]

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt (hunk 5 above)
Comment (Davran331122) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt:34]: I can, but I wanted to do not touch and moving app classes

File: app/src/main/java/org/p2p/wallet/user/repository/prices/impl/TokenPricesCoinGeckoRepository.kt (hunk 1 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/user/repository/prices/impl/TokenPricesCoinGeckoRepository.kt:63]: I Remove this methods from coingecko, so it's out of date

File: app/src/main/java/org/p2p/wallet/AppModule.kt [hunk 22]
```diff
@@ -91,10 +91,9 @@ object AppModule {
                 SendModule.create(),
                 TransactionManagerModule.create(),
                 SolendModule.create(),
-                EthereumModule.create(),
                 SellModule.create(),
                 ClaimModule.create()
-            )
+            ) + EthereumKitService.getEthereumKitModules()
```
[the remark this answers is not in the document]
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/AppModule.kt:96]: no, cause it's input list of modules, so we have one from app and another from eth kit, so I join them by + command

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/EthereumNetworkModule.kt (hunk 6 above)
Comment (Davran331122) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/EthereumNetworkModule.kt:27]: Provide named scope for this dependency

File: ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt (hunk 21 above)
Comment (Davran331122) [reply ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt:43]: I have not this dependencies in my module x D

**Pull request merged**